src/particle_filter/src/particleFilter6DOF.cpp

```diff
@@ -0,0 +1,359 @@
+
+particleFilter::particleFilter(int n_particles, cspace b_init[2],
+			       double Xstd_ob, double Xstd_tran,
+			       double Xstd_scatter, double R)
+  : N(n_particles), Xstd_ob(Xstd_ob), Xstd_tran(Xstd_tran),
+    Xstd_scatter(Xstd_scatter), R(R), firstObs(true)
+{
+  memcpy(b_Xprior, b_init, 2 * sizeof(cspace));
+  memcpy(b_Xpre, b_Xprior, 2 * sizeof(cspace));
+
+  X = new cspace[N];
+  bzero(X, N*sizeof(cspace));
+  X0 = new cspace[N];
+  bzero(X0, N*sizeof(cspace));
+
+  createParticles(X0, b_Xprior, N);
+
+  X_1 = new cspace[N];
+  W = new double[N];
+}
+
+void particleFilter::addObservation(double obs[3], double cube[3], distanceTransform *dist_transform, int idx_obs)
+{
+  std::default_random_engine generator;
+  normal_distribution<double> dist2(0, Xstd_scatter);
+
+  if (!firstObs) {
+    bzero(b_Xpre, 2 * sizeof(cspace));
+    for (int k = 0; k < cdim; k++) {
+      for (int j = 0; j < N; j++) {
+	X0[j][k] += dist2(generator);
+	b_Xpre[0][k] += X0[j][k];
+      }
+      b_Xpre[0][k] /= N;
+      for (int j = 0; j < N; j++) {
+	b_Xpre[1][k] += SQ(X0[j][k] - b_Xpre[0][k]);
+      }
+      b_Xpre[1][k] = sqrt(b_Xpre[1][k] / N);
+    }
+  }
+  bool iffar = updateParticles(X_1, X0, X, obs, cube, idx_obs, dist_transform, N, R, Xstd_ob, Xstd_tran);
+  if (firstObs)
+    {
+      firstObs = false;
+      memcpy(X0, X, N*sizeof(cspace));
+      memcpy(X_1, X, N*sizeof(cspace));
+    }
+  else if (iffar == true)
+    {
+      memcpy(X0, X_1, N*sizeof(cspace));
+    }
+  //calcWeight(W, N, Xstd_tran, X0, X);
+  memcpy(X_1, X0, N*sizeof(cspace));
+  //resampleParticles(X0, X, W, N);
+  memcpy(X0, X, N*sizeof(cspace));
+
+  for (int k = 0; k < cdim; k++) {
+    X_est[k] = 0;
+    for (int j = 0; j < N; j++) {
+      X_est[k] += X0[j][k];
+    }
+    X_est[k] /= N;
+  }
+  X_est_stat[1] = 0;
+  for (int j = 0; j < N; j++)
+    {
+      for (int k = 0; k < cdim; k++) {
+	X_est_stat[1] += SQ(X0[j][k] - X_est[k]);
+      }
+    }
+  X_est_stat[1] = sqrt(X_est_stat[1] / N);
+
+  if (X_est_stat[1] < 0.005 && (abs(X_est[0] - b_Xpre[0][0])>0.001 ||
+				abs(X_est[1] - b_Xpre[0][1])>0.001 ||
+				abs(X_est[2] - b_Xpre[0][2])>0.001 ||
+				abs(X_est[3] - b_Xpre[0][3])>0.001 ||
+				abs(X_est[4] - b_Xpre[0][4]) > 0.001 ||
+				abs(X_est[5] - b_Xpre[0][5]) > 0.001))
+    Xstd_scatter = 0.01;
+  else
+    Xstd_scatter = 0.0001;
+}
+
+void particleFilter::createParticles(cspace *X, cspace b_Xprior[2],
+				     int n_particles)
+{
+  random_device rd;
+  mt19937 e2(rd());
+  normal_distribution<double> dist(0, 1);
+  int cdim = sizeof(cspace) / sizeof(double);
+  for (int i = 0; i < n_particles; i++){
+    for (int j = 0; j < cdim; j++){
+      X[i][j] = b_Xprior[0][j] + b_Xprior[1][j] * (dist(e2));
+    }
+  }
+}
+
+bool particleFilter::updateParticles(cspace *X_1, cspace *X0, cspace *X, double cur_M[3], 
+				     double cube[3], int idx_Measure, distanceTransform *dist_transform, int n_particles,
```

Comment:
As mentioned above, use 
&variable 
to pass by reference. No need for pointers